---
title: "mirna_analysis"
author: "Barry"
date: "5/4/2022"
output: html_document
---

```{R}
library(DESeq2)
library(ggpubr)
library(biomaRt)

get_upregulated <- function(df){

	key <- intersect(rownames(df)[which(df$log2FoldChange>=1)], rownames(df)[which(df$padj<=0.05)])

    results <- as.data.frame((df)[which(rownames(df) %in% key),])
	return(results)
}

get_downregulated <- function(df){

  	key <- intersect(rownames(df)[which(df$log2FoldChange<=-1)], rownames(df)[which(df$padj<=0.05)])

  	results <- as.data.frame((df)[which(rownames(df) %in% key),])
  	return(results)
}

```

#######################################
# 1. miRNA
#######################################

Drop the metastatic sample here
```{R}
load("/data/TCGA-PRAD/mirna_meta/mirna.RData")

mirna_counts <- mirna_counts[ , -which(names(mirna_counts) %in% c("TCGA-V1-A9O5-06A"))]
mirna_meta <- mirna_meta[!(mirna_meta$Sample.ID == "TCGA-V1-A9O5-06A"),]
mirna_meta$Sample.Type <- ifelse(mirna_meta$Sample.Type == "Solid Tissue Normal", "Normal", mirna_meta$Sample.Type)
mirna_meta$Sample.Type <- ifelse(mirna_meta$Sample.Type == "Primary Tumor", "Tumor", mirna_meta$Sample.Type)
```


```{R, message=F}
library(DESeq2)
library(IHW)
library(apeglm)

dds <- DESeqDataSetFromMatrix(mirna_counts, mirna_meta, design = ~ Sample.Type)
dds$condition <- relevel(dds$Sample.Type, ref = "Normal")
dds <- DESeq(dds)
resultsNames(dds)
```

# EDA

```{R}
mirna_vst <- varianceStabilizingTransformation(dds, blind=F)
mirna_vst <- assay(mirna_vst)
PCA <- PCAtools::pca(mirna_vst, metadata = mirna_meta, removeVar = 0.1)

png("/data/TCGA-PRAD/mirna_meta/PCA_corr.png", width = 960, height = 480, units = "px")
PCAtools::eigencorplot(PCA, components = PCAtools::getComponents(PCA, 1:10),
    metavars = c('age_at_index','race','Sample.Type','ajcc_pathologic_t'),
    col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
    cexCorval = 1.2, fontCorval = 2, posLab = 'all', rotLabX = 45,
    scale = TRUE,
    main = bquote(PC ~ Pearson ~ r^2 ~ clinical ~ correlates),
    plotRsquared = TRUE,
    corFUN = 'pearson',
    corUSE = 'pairwise.complete.obs',
    corMultipleTestCorrection = 'BH',
    signifSymbols = c('****', '***', '**', '*', ''),
    signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1))
dev.off()

png("/data/TCGA-PRAD/mirna_meta/PCA_biplot.png", width = 960, height = 600, units="px")
PCAtools::biplot(PCA, x="PC1", y="PC2", colby = "Sample.Type", colkey = c(Normal = "forestgreen", Tumor = "blue"),
                 drawConnectors = FALSE, ellipse = T, ellipseFill = T, ellipseAlpha = 0.25, lab = NULL,
                 title = "TCGA-PRAD variance stabilized miRNA PCA Plot", gridlines.major = T, gridlines.minor = T,
                 legendPosition = "right")
dev.off()
```

# RESULTS
```{R}
res <- results(dds, filterFun=ihw, alpha=0.05, c("Sample.Type", "Tumor", "Normal"))
lfc_res <- lfcShrink(dds=dds, res=res, coef=2, type="apeglm")
res_df <- as.data.frame(lfc_res)
#plotCounts(dds, "hsa-mir-5704", intgroup = "Sample.Type") # sanity check of fold change direction.
up <- get_upregulated(res_df)
down <- get_downregulated(res_df)

de_mirs <- rbind(up,down)
```

# heatmap looks bad.

```{R, fig.height = 5}
mat <- as.data.frame(mirna_vst[rownames(de_mirs),])
mat$var <- apply(mat,1,var)
mat <- mat[order(mat$var, decreasing = T),]
mat <- as.matrix(mat[1:5,1:550])
mat <- t(mat)
mat <- scale(mat, center=T, scale = T)
mat <- t(mat)

annotation_col <- data.frame(Sample_Type = factor(mirna_meta$Sample.Type))
rownames(annotation_col) = colnames(mat)
colors <- c("forestgreen", "blue")
names(colors) <- c("Normal", "Tumor")
ann_col <- list(Sample_Type = colors)

#png("/data/TCGA-PRAD/mirna_meta/de_heat.png", width=500, height = 800, units="mm", res = 500)
pheatmap::pheatmap(mat, main = "Differentially Expressed miRNAs TCGA-PRAD",
         #col = hcl.colors(50, "Spectral",rev=T),
         cluster_cols = TRUE,
         show_rownames=TRUE, show_colnames = FALSE,
         annotation_col = annotation_col,
         annotation_colors = ann_col)
#dev.off()
```

# Annotate with genomic locations for Circos etc.

```{R}
library(biomaRt)
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")

info <- getBM(attributes=c("mirbase_id",
                               "ensembl_gene_id_version",
                               "chromosome_name",
                               "start_position",
                               "end_position",
                               "strand"),
                  filters = c("mirbase_id"),
                  values = rownames(de_mirs),
                  mart = mart,
                  useCache=FALSE)

info$strand <- gsub("-1", "-", info$strand)
info$strand <- gsub("1", "+", info$strand)

## Alt chroms account for duplicated mir IDS. 
info <- info[!grepl("CHR_", info$chromosome_name),]

de_mirs$mirbase_id <- rownames(de_mirs)
de_mirs <- merge(de_mirs, info, by = "mirbase_id")
de_mirs <- de_mirs[,c(1,3,5:11)]
write.table(de_mirs, "/data/TCGA-PRAD/mirna_res/de_mirs.txt", row.names = F, sep="\t", quote=F)
```

#######################################
# 2. mRNA
#######################################

```{R}
load("/data/TCGA-PRAD/mrna_meta/mrna.RData")

mrna_counts <- mrna_counts[ , -which(names(mrna_counts) %in% c("TCGA-V1-A9O5-06A"))]
mrna_meta <- mrna_meta[!(mrna_meta$Sample.ID == "TCGA-V1-A9O5-06A"),]
mrna_meta$Sample.Type <- ifelse(mrna_meta$Sample.Type == "Solid Tissue Normal", "Normal", mrna_meta$Sample.Type)
mrna_meta$Sample.Type <- ifelse(mrna_meta$Sample.Type == "Primary Tumor", "Tumor", mrna_meta$Sample.Type)
```

```{R}
dds <- DESeqDataSetFromMatrix(mrna_counts, mrna_meta, design = ~ Sample.Type)
dds$condition <- relevel(dds$Sample.Type, ref = "Normal")
dds <- DESeq(dds)
resultsNames(dds)
```

# EDA

```{R}
mrna_vst <- varianceStabilizingTransformation(dds, blind=F)
mrna_vst <- assay(mrna_vst)
PCA <- PCAtools::pca(mrna_vst, metadata = mrna_meta, removeVar = 0.1)

png("/data/TCGA-PRAD/mrna_meta/PCA_corr.png", width = 960, height = 480, units = "px")
PCAtools::eigencorplot(PCA, components = PCAtools::getComponents(PCA, 1:10),
    metavars = c('age_at_index','race','Sample.Type','ajcc_pathologic_t'),
    col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
    cexCorval = 1.2, fontCorval = 2, posLab = 'all', rotLabX = 45,
    scale = TRUE,
    main = bquote(PC ~ Pearson ~ r^2 ~ clinical ~ correlates),
    plotRsquared = TRUE,
    corFUN = 'pearson',
    corUSE = 'pairwise.complete.obs',
    corMultipleTestCorrection = 'BH',
    signifSymbols = c('****', '***', '**', '*', ''),
    signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1))
dev.off()

png("/data/TCGA-PRAD/mrna_meta/PCA_biplot.png", width = 960, height = 600, units="px")
PCAtools::biplot(PCA, x="PC1", y="PC3", colby = "Sample.Type", colkey = c(Normal = "forestgreen", Tumor = "blue"),
                 drawConnectors = FALSE, ellipse = T, ellipseFill = T, ellipseAlpha = 0.25, lab = NULL,
                 title = "TCGA-PRAD variance stabilized mRNA PCA Plot", gridlines.major = T, gridlines.minor = T,
                 legendPosition = "right")
dev.off()
```

```{R}
res <- results(dds, filterFun=ihw, alpha=0.05, c("Sample.Type", "Tumor", "Normal"))
lfc_res <- lfcShrink(dds=dds, res=res, coef=2, type="apeglm")
res_df <- as.data.frame(lfc_res)
#plotCounts(dds, "ENSG00000122133.17", intgroup = "Sample.Type") # sanity check of fold change direction.
up <- get_upregulated(res_df)
down <- get_downregulated(res_df)

de_genes <- rbind(up,down)
```

```{R}
library(biomaRt)
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")

rownames(de_genes) <- sub("\\..*", "", rownames(de_genes))

info <- getBM(attributes=c("hgnc_symbol",
                           "ensembl_gene_id",
                           "ensembl_gene_id_version",
                               "chromosome_name",
                               "start_position",
                               "end_position",
                               "strand"),
                  filters = c("ensembl_gene_id"),
                  values = rownames(de_genes),
                  mart = mart,
                  useCache=FALSE)

info$strand <- gsub("-1", "-", info$strand)
info$strand <- gsub("1", "+", info$strand)

# missing 6, cant retrieve info (crucial for CPG overlaps) so must settle and discard these. 
de_genes$ensembl_gene_id <- rownames(de_genes)
#info <- subset(info, select=-c(ensembl_gene_id))
info$strand <- gsub("-1", "-", info$strand)
info$strand <- gsub("1", "+", info$strand)


de_genes <- merge(de_genes, info, by = "ensembl_gene_id")
de_genes <- de_genes[,c(7,3,5,6,8:12)]
write.table(de_genes, "/data/TCGA-PRAD/mrna_res/de_genes.txt", row.names = F, sep="\t", quote=F)
```

# Methylaion DE

```{R}
load("/data/TCGA-PRAD/methylation_meta/mSetSq_filt.RData")
methyl_meta$Sample_Type <- ifelse(methyl_meta$Sample_Type == "Primary Tumor", "Tumor", methyl_meta$Sample_Type)
# bizarre introduction of 3's here instead of methyl_meta$Sample_Type
methyl_meta$Sample_Type <- ifelse(methyl_meta$Sample_Type == "3", "Normal", methyl_meta$Sample_Type)
```


# EDA
```{R}
pal <- RColorBrewer::brewer.pal(2, "Dark2")
png("/data/TCGA-PRAD/methylation_meta/PCA_biplot.png")
limma::plotMDS(mVals, top=10000, gene.selection="common", pch = 16,
        col=pal[factor(methyl_meta$Sample_Type)])
legend("bottomright", legend=levels(factor(methyl_meta$Sample_Type)), text.col=pal,
       bg="white", cex=0.7)
dev.off()
```

# LIMMA
```{R}
# this is the factor of interest
SampleType <- factor(methyl_meta$Sample_Type)

# use the above to create a design matrix
design <- model.matrix(~0+Sample_Type, data=methyl_meta)
colnames(design) <- c("Normal", "Tumor")
 
# fit the linear model 
fit <- limma::lmFit(mVals, design)
# create a contrast matrix for specific comparisons
contMatrix <- limma::makeContrasts(Tumor_vs_Normal = Tumor-Normal, levels=design)

fit2 <- limma::contrasts.fit(fit, contMatrix)
fit2 <- limma::eBayes(fit2)
summary(limma::decideTests(fit2))
```

```{R}
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
ann450k_ranges <- makeGRangesFromDataFrame(ann450k[,1:4],
                              keep.extra.columns=T,
                              ignore.strand=FALSE,
                              seqinfo=NULL,
                              seqnames.field=c("seqnames", "seqname",
                                               "chromosome", "chrom",
                                               "chr", "chromosome_name",
                                               "seqid"),
                              start.field="pos",
                              end.field="pos",
                              strand.field="strand",
                              starts.in.df.are.0based=FALSE)
chain=rtracklayer::import.chain("/data/TCGA-PRAD/methylation_meta/hg19ToHg38.over.chain")
hg38loc <- rtracklayer::liftOver(ann450k_ranges,chain)
hg38locDF <- data.frame(hg38loc)
rownames(hg38locDF) <- hg38locDF$group_name
pos38<-start(unlist(hg38loc))
ann450k_ranges=data.frame(ann450k,"pos.hg19"=ann450k$pos)
ann450k_ranges$pos=rep(NA,dim(ann450k_ranges)[1])
ann450k_ranges[hg38locDF$Name, "pos"]=hg38locDF[,"start"]

ann450kSub <- ann450k_ranges[match(rownames(mVals),ann450k_ranges$Name),
                      c(1:4,12:19,24:ncol(ann450k_ranges))]

DMPs <- limma::topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)

# sanity check for limma design below (tumor is up )

par(mfrow=c(2,2))
sapply(rownames(DMPs)[1:4], function(cpg){
  plotCpg(mVals, cpg=cpg, pheno=methyl_meta$Sample_Type, ylab = "M values")
})

# quote = T important to mark empty cells properly for readign back to R
write.table(DMPs, "/data/TCGA-PRAD/methylation_res/DMPs.txt", sep="\t", row.names = T, quote=T)

## apply filtering
DMP_filt <- subset(DMPs, DMPs$adj.P.Val < 0.05)
up <- subset(DMP_filt, DMP_filt$logFC > 1)
down <- subset(DMP_filt, DMP_filt$logFC < -1)
DMP_filt <- rbind(up,down)

write.table(DMP_filt, "/data/TCGA-PRAD/methylation_res/dmp_filt.txt", sep="\t", row.names = T, quote=T)
```

