---
title: "mirna_analysis"
author: "Barry"
date: "5/4/2022"
output: html_document
---

```{R}
library(DESeq2)
library(ggpubr)
library(biomaRt)

get_upregulated <- function(df){

	key <- intersect(rownames(df)[which(df$log2FoldChange>=0.5)], rownames(df)[which(df$padj<=0.05)])

    results <- as.data.frame((df)[which(rownames(df) %in% key),])
	return(results)
}

get_downregulated <- function(df){

  	key <- intersect(rownames(df)[which(df$log2FoldChange<=-0.5)], rownames(df)[which(df$padj<=0.05)])

  	results <- as.data.frame((df)[which(rownames(df) %in% key),])
  	return(results)
}

```

# stage data and perform EDA.

Drop the metastatic sample here
```{R}
load("/data/TCGA-PRAD/mirna_meta/mirna.RData")

mirna_counts <- mirna_counts[ , -which(names(mirna_counts) %in% c("TCGA-V1-A9O5-06A"))]
mirna_meta <- mirna_meta[!(mirna_meta$Sample.ID == "TCGA-V1-A9O5-06A"),]
mirna_meta$Sample.Type <- ifelse(mirna_meta$Sample.Type == "Solid Tissue Normal", "Normal", mirna_meta$Sample.Type)
mirna_meta$Sample.Type <- ifelse(mirna_meta$Sample.Type == "Primary Tumor", "Tumor", mirna_meta$Sample.Type)
```


```{R}
library(DESeq2)
library(IHW)
library(apeglm)

dds <- DESeqDataSetFromMatrix(mirna_counts, mirna_meta, design = ~ Sample.Type)
dds$condition <- relevel(dds$Sample.Type, ref = "Normal")
dds <- DESeq(dds)
resultsNames(dds)
```

# EDA

```{R}
mirna_vst <- vst(dds, blind=F)
PCA <- PCAtools::pca(mirna_vst, metadata = meta, remove)

```

# RESULTS
```{R}
res <- results(dds, filterFun=ihw, alpha=0.05, c("Sample.Type", "Tumor", "Normal"))
lfc_res <- lfcShrink(dds=dds, res=res, coef=2, type="apeglm")
res_df <- as.data.frame(lfc_res)
#plotCounts(dds, "hsa-mir-5704", intgroup = "Sample.Type") # sanity check of fold change direction.
up <- get_upregulated(res_df)
down <- get_downregulated(res_df)

de_mirs <- rbind(up,down)
```

# Annotate with genomic locations for Circos etc.

```{R}
library(biomaRt)
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")

info <- getBM(attributes=c("mirbase_id",
                               "ensembl_gene_id_version",
                               "chromosome_name",
                               "start_position",
                               "end_position",
                               "strand"),
                  filters = c("mirbase_id"),
                  values = rownames(de_mirs),
                  mart = mart,
                  useCache=FALSE)

info$strand <- gsub("-1", "-", info$strand)
info$strand <- gsub("1", "+", info$strand)

## Alt chroms account for duplicated mir IDS. 
info <- info[!grepl("CHR_", info$chromosome_name),]

de_mirs$mirbase_id <- rownames(de_mirs)
de_mirs <- merge(de_mirs, info, by = "mirbase_id")
de_mirs <- de_mirs[,c(1,3,5:11)]
write.table(de_mirs, "/data/TCGA-PRAD/mirna_res/de_mirs.txt", row.names = F, sep="\t", quote=F)
```

```{R}
load("/data/TCGA-PRAD/mrna_meta/mrna.RData")

mrna_counts <- mrna_counts[ , -which(names(mrna_counts) %in% c("TCGA-V1-A9O5-06A"))]
mrna_meta <- mrna_meta[!(mrna_meta$Sample.ID == "TCGA-V1-A9O5-06A"),]
mrna_meta$Sample.Type <- ifelse(mrna_meta$Sample.Type == "Solid Tissue Normal", "Normal", mrna_meta$Sample.Type)
mrna_meta$Sample.Type <- ifelse(mrna_meta$Sample.Type == "Primary Tumor", "Tumor", mrna_meta$Sample.Type)
```

```{R}
dds <- DESeqDataSetFromMatrix(mrna_counts, mrna_meta, design = ~ Sample.Type)
dds$condition <- relevel(dds$Sample.Type, ref = "Normal")
dds <- DESeq(dds)
resultsNames(dds)

res <- results(dds, filterFun=ihw, alpha=0.05, c("Sample.Type", "Tumor", "Normal"))
lfc_res <- lfcShrink(dds=dds, res=res, coef=2, type="apeglm")
res_df <- as.data.frame(lfc_res)
#plotCounts(dds, "ENSG00000122133.17", intgroup = "Sample.Type") # sanity check of fold change direction.
up <- get_upregulated(res_df)
down <- get_downregulated(res_df)

de_genes <- rbind(up,down)
```

```{R}
library(biomaRt)
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")

rownames(de_genes) <- sub("\\..*", "", rownames(de_genes))

info <- getBM(attributes=c("hgnc_symbol",
                           "ensembl_gene_id",
                           "ensembl_gene_id_version",
                               "chromosome_name",
                               "start_position",
                               "end_position",
                               "strand"),
                  filters = c("ensembl_gene_id"),
                  values = rownames(de_genes),
                  mart = mart,
                  useCache=FALSE)

info$strand <- gsub("-1", "-", info$strand)
info$strand <- gsub("1", "+", info$strand)

# missing 6, cant retrieve info (crucial for CPG overlaps) so must settle and discard these. 
de_genes$ensembl_gene_id <- rownames(de_genes)
#info <- subset(info, select=-c(ensembl_gene_id))
info$strand <- gsub("-1", "-", info$strand)
info$strand <- gsub("1", "+", info$strand)


de_genes <- merge(de_genes, info, by = "ensembl_gene_id")
de_genes <- de_genes[,c(7,3,5,6,8:12)]
write.table(de_genes, "/data/TCGA-PRAD/mrna_res/de_genes.txt", row.names = F, sep="\t", quote=F)
```

# Methylaion Filtering, DE

```{R}
load("/data/TCGA-PRAD/methylation_meta/methyl.RData")

methyl_meta <- methyl_meta[!(methyl_meta$Sample_ID == "TCGA-V1-A9O5-06A"),]

keep <- !(sampleNames(mSetSq) == "TCGA-V1-A9O5-06A")
mSetSq <- mSetSq[,keep]
mSetSq

mSetSq <- dropLociWithSnps(mSetSq)
mSetSq

reactive_probes <- read.csv("/data/TCGA-PRAD/methylation_meta/cross_reactive_probes.csv", sep=",", header=T)
keep <- !(featureNames(mSetSq) %in% reactive_probes$TargetID)
mSetSq <- mSetSq[keep,] 
mSetSq
```

```{r}
rm(reactive_probes, keep)
mVals <- getM(mSetSq)

bVals <- getBeta(mSetSq)

mSetSeq_flt <- mSetSq
rm(mSetSeq)

save.image("/data/TCGA-PRAD/methylation_meta/mSetSq_filt.RData")
```

```{R}
methyl_meta$Sample_Type <- ifelse(methyl_meta$Sample_Type == "Primary Tumor", "Tumor", methyl_meta$Sample_Type)
methyl_meta$Sample_Type <- ifelse(methyl_meta$Sample_Type == "Solid Tissue Normal", "Normal", methyl_meta$Sample_Type)

# this is the factor of interest
SampleType <- factor(methyl_meta$Sample_Type)

# use the above to create a design matrix
design <- model.matrix(~0+Sample_Type, data=methyl_meta)
colnames(design) <- c("Normal", "Tumor")
 
# fit the linear model 
fit <- limma::lmFit(mVals, design)
# create a contrast matrix for specific comparisons
contMatrix <- limma::makeContrasts(Tumor_vs_Normal = Tumor-Normal, levels=design)

fit2 <- limma::contrasts.fit(fit, contMatrix)
fit2 <- limma::eBayes(fit2)
summary(limma::decideTests(fit2))
```

```{R}
ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)

ann450kSub <- ann450k[match(rownames(mVals),ann450k$Name),
                      c(1:4,12:19,24:ncol(ann450k))]
DMPs <- topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)

# sanity check for limma design below (tumor is up )

par(mfrow=c(2,2))
sapply(rownames(DMPs)[1:4], function(cpg){
  plotCpg(bVals, cpg=cpg, pheno=methyl_meta$Sample_Type, ylab = "Beta values")
})

# quote = T important to mark empty cells properly for readign back to R
write.table(DMPs, "/data/TCGA-PRAD/methylation_res/DMPs.txt", sep="\t", row.names = T, quote=T)
```

